dashboard/app/models/levels/level.rb

```diff
@@ -43,6 +43,12 @@ class Level < ActiveRecord::Base
   has_many :levels_child_levels, class_name: 'ParentLevelsChildLevel', foreign_key: :parent_level_id
   has_many :child_levels, through: :levels_child_levels, inverse_of: :parent_levels
 
+  has_one :levels_contained_level, -> {where(kind: 'contained')}, class_name: 'ParentLevelsChildLevel', foreign_key: :parent_level_id
```

Comment:
Is this always a 1:1 thing? If so does this need to use the join table at all? Couldn't it use class_name 'Level' instead?

(Does the same also apply for `template` as well?)

Reply:
> Is this always a 1:1 thing?

I'm not sure. it is at most many to one, i.e. each level only needs one contained level, but there's nothing stopping multiple levels from having the same contained level. I can check whether it is in fact only one-to-one if we think that would affect the decision (see below regarding template levels).

>  If so does this need to use the join table at all?

Not necessarily -- the other approach in https://github.com/code-dot-org/code-dot-org/pull/34585 avoids using the join table entirely by using a `contained_level_id` column on the `levels` table, and the point of showing both options is to compare the two approaches.

> Couldn't it use class_name 'Level' instead?

How do you mean -- using the `contained_level_id` approach described in #34585, or a different way?

> (Does the same also apply for template as well?)

I was going to say "yes", but template levels do differ in that they are definitely shared by multiple levels. So, even if we require that contained levels are one-to-one and that leads to a simpler implementation, we won't be able to apply the same simplification to template levels. However, if we assume both relationships are many-to-one then we can implement them similarly.

Reply:
> Not necessarily -- the other approach in #34585 avoids using the join table entirely by using a contained_level_id column on the levels table, and the point of showing both options is to compare the two approaches.

Yea, sorry, I wasn't thinking clearly. #34585 is what I was thinking about here.

A line of questioning might be: will there be any other differences between the two approaches in the code? From these two PRs, it seems like the only difference is in `all_depending_levels`, which other parts of the code like `all_scripts` can call.

Reply:
I ask because if there aren't really other differences in how the rest of the code base will need to interact with this, it almost seems like it doesn't matter. Is that too naive?

Reply:
> A better line of questioning might be: will there be any other differences between the two approaches in the code? From these two PRs, it seems like the only difference is in `all_depending_levels`, which other parts of the code like `all_scripts` can call.

I've updated https://github.com/code-dot-org/code-dot-org/pull/34585 to include the logic for template levels, in hopes to fully illustrate any complexity that might appear with that approach. The way I've arranged it, the complexity appears to be limited to the `all_parent_levels` and `all_child_levels` methods on that PR.

Reply:
> I ask because if there aren't really other differences in how the rest of the code base will need to interact with this, it almost seems like it doesn't matter. Is that too naive?

I think you're right -- the choice I'm framing in these two PRs won't really affect the rest of the system. The only thing I can think of is that this PR might make it slightly easier for the data team to write SQL queries, which is a valid concern, but I also don't want to over-index on that.

Reply:
FWIW @uponthesun I think I've arrived at the same conclusion -- I had a stronger opinion initially, but now that I see both approaches I would be fine with either. I want to wait for Elijah's thoughts (just tagged below) before proceeding because he also had an initial opinion. If no one feels strongly, I'm happy to arbitrate.